src/opt.c

```diff
@@ -248,10 +249,18 @@ void fill_segment(const argon2_instance_t *instance,
             pseudo_rand = address_block.v[i % ARGON2_ADDRESSES_IN_BLOCK];
         } else {
             pseudo_rand = instance->memory[prev_offset].v[0];
+            prev_offset = curr_offset;
         }
 
         /* 1.2.2 Computing the lane of the reference block */
-        ref_lane = ((pseudo_rand >> 32)) % instance->lanes;
+        if (lanes_reciprocal == 0) {
+          ref_lane = (pseudo_rand >> 32) & (instance->lanes - 1);
+        } else {
+          ref_lane = (pseudo_rand >> 32) - (((pseudo_rand >> 32) * lanes_reciprocal) >> 32) * instance->lanes;
+          if (ref_lane >= instance->lanes) {
+            ref_lane -= instance->lanes;
+          }
```

Comment:
As I understand, when running with data dependent addressing the `ref_lane >= instance->lane` check makes things worse by also introducing data dependent branching. Can we avoid that?

Reply:
> why at most one subtraction is enough

When you do `reciprocal = floor(1 / n * 2**32)` and `floor(x * reciprocal / 2**32)` (ie `floor(x * floor(1 / n * 2**32) / 2**32)`) where `x` is a 32 bit integer:
```
x / n
= x * 1 / n
= x * (1 / n * 2**32) / 2**32
< x * floor(1 / n * 2**32) / 2**32  // the thing
< x * (1 / n * 2**32 + 1) / 2**32
= (x / n * 2**32 + x) / 2**32
= x / n + x / 2**32
< x / n + 1  // note x < 2**32 thus x / 2**32 < 1
so
x / n < x * floor(1 / n * 2**32) / 2**32 < x / n + 1
thus
floor(x / n) ≤ floor(x * floor(1 / n * 2**32) / 2**32) ≤ floor(x / n) + 1
```

----

Sorry I've been a lazy piece of shit. I don't know what's wrong with me. I need to like do stuff. Anyway hopefully I didn't fuck up that proof since I'm quite drunk but sort of sobered up in the middle it. (I've proved it to myself a few times but I think this is the first time I've been like oh yeah that's like a real proof.)